File: language/runtime/apple/core/Sources/RuntimeHostAppleCore/Module/Contact/ContactRequests.swift
import Foundation

/// The Apple contact request surface for one runtime host embedder.
@MainActor
public protocol ContactRequests: AnyObject {
    /// List one page of contacts for one query.
    func listContacts(
        _ query: RuntimeHostContactQuery
    ) -> RuntimeHostContactPageResponse

    /// Search contacts for one query string and one query shape.
    func searchContacts(
        _ queryText: String,
        query: RuntimeHostContactQuery
    ) -> RuntimeHostContactPageResponse

    /// Read one contact by stable identifier.
    func readContact(
        id: String
    ) -> RuntimeHostContactResponse

    /// Create one contact and return its stable identifier.
    func createContact(
        _ draft: RuntimeHostContactDraft
    ) -> RuntimeHostContactCreateResponse

    /// Update one contact by stable identifier.
    func updateContact(
        id: String,
        draft: RuntimeHostContactDraft
    ) -> UInt32

    /// Delete one contact by stable identifier.
    func deleteContact(
        id: String
    ) -> UInt32
}

/// The explicit unsupported contact request surface for one Apple runtime host.
@MainActor
public final class UnsupportedContactRequests: ContactRequests {
    /// Create one unsupported contact request surface.
    public init() {}

    public func listContacts(
        _ query: RuntimeHostContactQuery
    ) -> RuntimeHostContactPageResponse {
        RuntimeHostContactPageResponse(status: hostStatusNotSupported)
    }

    public func searchContacts(
        _ queryText: String,
        query: RuntimeHostContactQuery
    ) -> RuntimeHostContactPageResponse {
        RuntimeHostContactPageResponse(status: hostStatusNotSupported)
    }

    public func readContact(
        id: String
    ) -> RuntimeHostContactResponse {
        RuntimeHostContactResponse(status: hostStatusNotSupported)
    }

    public func createContact(
        _ draft: RuntimeHostContactDraft
    ) -> RuntimeHostContactCreateResponse {
        RuntimeHostContactCreateResponse(status: hostStatusNotSupported)
    }

    public func updateContact(
        id: String,
        draft: RuntimeHostContactDraft
    ) -> UInt32 {
        hostStatusNotSupported
    }

    public func deleteContact(
        id: String
    ) -> UInt32 {
        hostStatusNotSupported
    }
}

File: language/runtime/android/kotlin/src/main/kotlin/dev/destack/runtime/android/module/contact/ContactRequests.kt
package dev.destack.runtime.android.module.contact

import dev.destack.runtime.android.core.hostStatusNotSupported

/**
 * The contact request surface for one Android runtime host.
 */
public interface ContactRequests {
    /**
     * List one page of contacts for one query.
     */
    public fun listContacts(
        query: RuntimeHostContactQuery,
    ): RuntimeHostContactPageResponse

    /**
     * Search contacts for one query string and one query shape.
     */
    public fun searchContacts(
        queryText: String,
        query: RuntimeHostContactQuery,
    ): RuntimeHostContactPageResponse

    /**
     * Read one contact by stable identifier.
     */
    public fun readContact(
        id: String,
    ): RuntimeHostContactResponse

    /**
     * Create one contact and return its stable identifier.
     */
    public fun createContact(
        draft: RuntimeHostContactDraft,
    ): RuntimeHostContactCreateResponse

    /**
     * Update one contact by stable identifier.
     */
    public fun updateContact(
        id: String,
        draft: RuntimeHostContactDraft,
    ): Int

    /**
     * Delete one contact by stable identifier.
     */
    public fun deleteContact(
        id: String,
    ): Int
}

/**
 * The explicit unsupported contact request surface for one Android runtime host.
 */
public object UnsupportedContactRequests : ContactRequests {
    override fun listContacts(
        query: RuntimeHostContactQuery,
    ): RuntimeHostContactPageResponse {
        return RuntimeHostContactPageResponse(status = hostStatusNotSupported)
    }

    override fun searchContacts(
        queryText: String,
        query: RuntimeHostContactQuery,
    ): RuntimeHostContactPageResponse {
        return RuntimeHostContactPageResponse(status = hostStatusNotSupported)
    }

    override fun readContact(
        id: String,
    ): RuntimeHostContactResponse {
        return RuntimeHostContactResponse(status = hostStatusNotSupported)
    }

    override fun createContact(
        draft: RuntimeHostContactDraft,
    ): RuntimeHostContactCreateResponse {
        return RuntimeHostContactCreateResponse(status = hostStatusNotSupported)
    }

    override fun updateContact(
        id: String,
        draft: RuntimeHostContactDraft,
    ): Int {
        return hostStatusNotSupported
    }

    override fun deleteContact(
        id: String,
    ): Int {
        return hostStatusNotSupported
    }
}
